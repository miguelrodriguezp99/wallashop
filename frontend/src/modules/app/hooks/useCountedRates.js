import { useEffect, useState } from "react";
import { getNRates } from "../../../backend/postService";

import "../styles/posts.css";

export default function useCountedRates(id) {
	const [nrates, setNRates] = useState({});

	const onSuccessRates = (countedRates) => {
		setNRates(countedRates);
	};

	const onErrorsRates = () => {
		console.log("Error al obtener los rates.");
	};

	useEffect(() => {
		getNRates(id, onSuccessRates, onErrorsRates);
	}, [id]);

	return { nrates, setNRates };
}
