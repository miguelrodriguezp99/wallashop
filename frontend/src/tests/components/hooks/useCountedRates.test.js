import { renderHook } from '@testing-library/react';
import useCountedRates from '../../../modules/app/hooks/useCountedRates'; 
import { act } from 'react-dom/test-utils';

//Se está mockeando (simular una función) getNrates para asi devuelva datos simulados
//En este caso se devuelve un objeto con un total de 10 rates y un average de 5 (por ejemplo)
jest.mock('../../../backend/postService', () => ({
    getNRates: jest.fn().mockResolvedValue({
      totalRates: 10,
      averageRate: 5,
    }),
}));
  
const { getNRates } = require('../../../backend/postService');
  
describe('useCountedRates', () => {
    it('matches snapshot', async () => {
        const id = 123;
        //Se usa renderHook para renderizar el hook (useCountedRates)
        //Se mete un id como argumento
        const { result } = renderHook(() => useCountedRates(id));

        //Aqui se comprueba que se ha llamado a getNRates con el id y las dos funciones que se pasan como argumento
        // (una para manejar el exito y otra para manejar el error)
        expect(getNRates).toHaveBeenCalledWith(id, expect.any(Function), expect.any(Function));

        //Se usa act para esperar a que se resuelva la promesa simulada "setTimeout"
        //act se usa para "sincronizar" las acciones que causan las actalizaciones en el arbol de componentes
        //react espera que todas las actualizaciones de estado se completen antes de continuar con el test
        //Por tanto se usa setTimeout para ejecutar la función despues de un retraso "0" en este caso
        //De esta forma se simula una operación asíncrona que se completa despues de que la pila de ejecución actual
        //se haya vaciado, lo que permite que react pueda procesar las actualizaciones pendientes.
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0)); 
        });

        //Aqui se comprueba si el resultado del hook concuerda con el snapshot
        expect(result.current).toMatchSnapshot();
    });
});


