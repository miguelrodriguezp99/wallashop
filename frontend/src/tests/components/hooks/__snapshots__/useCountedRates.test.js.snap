// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCountedRates matches snapshot 1`] = `
Object {
  "nrates": Object {},
  "setNRates": [Function],
}
`;
